File: Elixir/Poker.ex
defmodule Poker do
    def deal([a, b, c, d | table]) do
        playerOneHand = [a, c] ++ table
        playerTwoHand = [b, d] ++ table

        playerOneHand = Enum.sort(playerOneHand)
        playerTwoHand = Enum.sort(playerTwoHand)

        determineWinner(playerOneHand, playerTwoHand)
    end

    def determineWinner(playerOneHand, playerTwoHand) do
        cards = [ nil, 
            "1C", "2C", "3C", "4C", "5C", "6C", "7C", "8C", "9C", "10C", "11C", "12C", "13C",
            "1D", "2D", "3D", "4D", "5D", "6D", "7D", "8D", "9D", "10D", "11D", "12D", "13D",
		    "1H", "2H", "3H", "4H", "5H", "6H", "7H", "8H", "9H", "10H", "11H", "12H", "13H",
	        "1S", "2S", "3S", "4S", "5S", "6S", "7S", "8S", "9S", "10S", "11S", "12S", "13S" ]

        valueComboOne = evaluateHand(playerOneHand)
        valueComboTwo = evaluateHand(playerTwoHand)

        valueOne = List.first(valueComboOne)
        valueTwo = List.first(valueComboTwo)

        comboOne = List.last(valueComboOne)
        comboTwo = List.last(valueComboTwo)

        cond do
            valueOne > valueTwo -> Enum.map(comboOne, fn index -> Enum.at(cards, index) end)
            valueOne < valueTwo -> Enum.map(comboTwo, fn index -> Enum.at(cards, index) end)
            true -> Enum.map(kicker(valueOne, comboOne, comboTwo, playerOneHand, playerTwoHand), fn index -> Enum.at(cards, index) end)
        end
    end

    def evaluateHand(hand) do
        cond do
            combo = royalFlushCheck(hand)        -> [10, combo]
            combo = straightFlushCheck(hand)     -> [09, combo]
            combo = duplicateCheck(hand, 4, nil) -> [08, combo]
            combo = fullHouseCheck(hand)         -> [07, combo]
            combo = flushCheck(hand)             -> [06, combo]
            combo = straightCheck(hand)          -> [05, combo]
            combo = duplicateCheck(hand, 3, nil) -> [04, combo]
            combo = twoPairCheck(hand)           -> [03, combo]
            combo = duplicateCheck(hand, 2, nil) -> [02, combo]
            true                                 -> [01, getHighCard(hand)]
        end
    end

    def royalFlushCheck(hand) do
        cond do
            MapSet.subset?(MapSet.new([01, 10, 11, 12, 13]), MapSet.new(hand)) -> [01, 10, 11, 12, 13]
            MapSet.subset?(MapSet.new([14, 23, 24, 25, 26]), MapSet.new(hand)) -> [14, 23, 24, 25, 26]
            MapSet.subset?(MapSet.new([27, 36, 37, 38, 39]), MapSet.new(hand)) -> [27, 36, 37, 38, 39]
            MapSet.subset?(MapSet.new([40, 49, 50, 51, 52]), MapSet.new(hand)) -> [40, 49, 50, 51, 52]
            true -> nil
        end
    end

    def straightFlushCheck([ a, b, c, d, e, f, g | _ ]) do
        cond do
            (c + 1 == d) && (c + 2 == e) && (c + 3 == f) && (c + 4 == g) -> [c, d, e, f, g]
            (b + 1 == c) && (b + 2 == d) && (b + 3 == e) && (b + 4 == f) -> [b, c, d, e, f]
            (a + 1 == b) && (a + 2 == c) && (a + 3 == d) && (a + 4 == e) -> [a, b, c, d, e]
            true -> nil
        end
    end

    def fullHouseCheck(hand) do
        triple = duplicateCheck(hand, 3, nil)
        pair   = duplicateCheck(hand, 2, nil)

        cond do
            triple && pair -> triple ++ pair
            true -> nil
        end
    end

    def flushCheck(hand) do
        rankValues = %{"1": 14, "0": 13, "12": 12, "11": 11, "10": 10, "9": 9, "8": 8, "7": 7, "6": 6, "5": 5, "4": 4, "3": 3, "2": 2}
        sortedRanks = Enum.sort(hand, &(rankValues[String.to_atom(Integer.to_string(rem(&1, 13)))] <= rankValues[String.to_atom(Integer.to_string(rem(&2, 13)))]))

        cond do
            Enum.count(hand, fn rank -> rank >= 01 && rank <= 13 end) >= 5 -> Enum.take(Enum.filter(sortedRanks, fn rank -> rank >= 01 && rank <= 13 end), -5)
            Enum.count(hand, fn rank -> rank >= 14 && rank <= 26 end) >= 5 -> Enum.take(Enum.filter(sortedRanks, fn rank -> rank >= 14 && rank <= 26 end), -5)
            Enum.count(hand, fn rank -> rank >= 27 && rank <= 39 end) >= 5 -> Enum.take(Enum.filter(sortedRanks, fn rank -> rank >= 27 && rank <= 39 end), -5)
            Enum.count(hand, fn rank -> rank >= 40 && rank <= 52 end) >= 5 -> Enum.take(Enum.filter(sortedRanks, fn rank -> rank >= 40 && rank <= 52 end), -5)
            true -> nil
        end
    end

    def straightCheck(hand) do
        rankValues = %{ "1": 1, "0": 13, "12": 12, "11": 11, "10": 10, "9": 9, "8": 8, "7": 7, "6": 6, "5": 5, "4": 4, "3": 3, "2": 2 }
        sortedRanks = Enum.sort(hand, &(rankValues[String.to_atom(Integer.to_string(rem(&1, 13)))] <= rankValues[String.to_atom(Integer.to_string(rem(&2, 13)))]))
        noDuplicate = Enum.uniq_by(sortedRanks, fn rank -> rankValues[String.to_atom(Integer.to_string(rem(rank, 13)))] end)
        
        case length(noDuplicate) do
            5 -> getStraight(noDuplicate ++ [-1] ++ [-1], rankValues)
            6 -> getStraight(noDuplicate ++ [-1], rankValues)
            7 -> getStraight(noDuplicate, rankValues)
            true -> nil
        end
    end
    
    def getStraight([ a, b, c, d, e, f, g | _ ], rankValues) do
        aReduced = rankValues[String.to_atom(Integer.to_string(rem(a, 13)))]
        bReduced = rankValues[String.to_atom(Integer.to_string(rem(b, 13)))]
        cReduced = rankValues[String.to_atom(Integer.to_string(rem(c, 13)))]
        dReduced = rankValues[String.to_atom(Integer.to_string(rem(d, 13)))]
        eReduced = rankValues[String.to_atom(Integer.to_string(rem(e, 13)))]
        fReduced = rankValues[String.to_atom(Integer.to_string(rem(f, 13)))]
        gReduced = rankValues[String.to_atom(Integer.to_string(rem(g, 13)))]

        cond do
            (dReduced + 1 == eReduced) && (dReduced + 2 == fReduced) && (dReduced + 3 == gReduced) && (dReduced - aReduced == dReduced - 1) -> [d, e, f, g, a]
            (cReduced + 1 == dReduced) && (cReduced + 2 == eReduced) && (cReduced + 3 == fReduced) && (cReduced - aReduced == cReduced - 1) -> [c, d, e, f, a]
            (bReduced + 1 == cReduced) && (bReduced + 2 == dReduced) && (bReduced + 3 == eReduced) && (bReduced - aReduced == bReduced - 1) -> [b, c, d, e, a]
            (cReduced + 1 == dReduced) && (cReduced + 2 == eReduced) && (cReduced + 3 == fReduced) && (cReduced + 4 == gReduced) -> [c, d, e, f, g]
            (bReduced + 1 == cReduced) && (bReduced + 2 == dReduced) && (bReduced + 3 == eReduced) && (bReduced + 4 == fReduced) -> [b, c, d, e, f]
            (aReduced + 1 == bReduced) && (aReduced + 2 == cReduced) && (aReduced + 3 == dReduced) && (aReduced + 4 == eReduced) -> [a, b, c, d, e]
            true -> nil
        end
    end

    def duplicateCheck(hand, duplicateType, pair) do
        duplicates = for index <- [1, 0, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2] do
            cond do
                Enum.count(hand, fn rank -> rem(rank, 13) == index end) == duplicateType && pair != index -> Enum.filter(hand, fn rank -> rem(rank, 13) == index end)
                true -> nil
            end
        end
        getDuplicate(duplicates)
    end

    def getDuplicate([head | tail]), do: head || getDuplicate(tail)
    def getDuplicate([]), do: nil

    def twoPairCheck(hand) do
        firstPair = duplicateCheck(hand, 2, nil)

        if firstPair do
            secondPair = duplicateCheck(hand, 2, rem(List.first(firstPair), 13))
            cond do
                firstPair && secondPair -> firstPair ++ secondPair
                true -> nil
            end
        end
    end

    def getHighCard(hand) do
        rankValues = %{ "1": 14, "0": 13, "12": 12, "11": 11, "10": 10, "9": 9, "8": 8, "7": 7, "6": 6, "5": 5, "4": 4, "3": 3, "2": 2 }
        sortedRanks = Enum.sort(hand, &(rankValues[String.to_atom(Integer.to_string(rem(&1, 13)))] <= rankValues[String.to_atom(Integer.to_string(rem(&2, 13)))]))
        [List.last(sortedRanks)]
    end

    def kicker(value, comboOne, comboTwo, playerOneHand, playerTwoHand) do
        cond do
            value == 5 or value == 9 -> kickerRank(0.01, comboOne, comboTwo, playerOneHand, playerTwoHand)
            true                     -> kickerRank(0.14, comboOne, comboTwo, playerOneHand, playerTwoHand)
        end
    end

    def kickerRank(aceValue, comboOne, comboTwo, playerOneHand, playerTwoHand) do
        rankValues = %{ "1": aceValue, "0": 0.13, "12": 0.12, "11": 0.11, "10": 0.10, "9": 0.09, "8": 0.08, "7": 0.07, "6": 0.06, "5": 0.05, "4": 0.04, "3": 0.03, "2": 0.02 }

        kickerOne = Enum.map(comboOne, fn rank -> rankValues[String.to_atom(Integer.to_string(rem(rank, 13)))] end)
        kickerTwo = Enum.map(comboTwo, fn rank -> rankValues[String.to_atom(Integer.to_string(rem(rank, 13)))] end)

        cond do
            Enum.sum(kickerOne) > Enum.sum(kickerTwo) -> comboOne
            Enum.sum(kickerOne) < Enum.sum(kickerTwo) -> comboTwo
            true -> kickerWholeHand(comboOne, comboTwo, playerOneHand, playerTwoHand)
        end
    end

    def kickerWholeHand(comboOne, comboTwo, playerOneHand, playerTwoHand) do
        rankValues = %{"1": 14, "0": 13, "12": 12, "11": 11, "10": 10, "9": 9, "8": 8, "7": 7, "6": 6, "5": 5, "4": 4, "3": 3, "2": 2}
        sortedOne  = Enum.sort(playerOneHand, &(rankValues[String.to_atom(Integer.to_string(rem(&1, 13)))] >= rankValues[String.to_atom(Integer.to_string(rem(&2, 13)))]))
        sortedTwo  = Enum.sort(playerTwoHand, &(rankValues[String.to_atom(Integer.to_string(rem(&1, 13)))] >= rankValues[String.to_atom(Integer.to_string(rem(&2, 13)))]))
        reducedOne = Enum.map(sortedOne, fn rank -> rankValues[String.to_atom(Integer.to_string(rem(rank, 13)))] end)
        reducedTwo = Enum.map(sortedTwo, fn rank -> rankValues[String.to_atom(Integer.to_string(rem(rank, 13)))] end)
        
        cardWinner = for index <- [0, 1, 2, 3, 4, 5, 6] do
            cond do
                Enum.at(reducedOne, index) > Enum.at(reducedTwo, index) -> comboOne
                Enum.at(reducedOne, index) < Enum.at(reducedTwo, index) -> comboTwo
                true -> nil
            end
        end

        getCombo(cardWinner)
    end

    def getCombo([head | tail]), do: head || getCombo(tail)
    def getCombo([]), do: nil
end

# 164 Lines

File: Haskell/Poker.hs
-- Tran Kelvin Lim (500769478)
-- Section 01
-- Brian Bao Nguyen (500818362)
-- Section 05

module Poker where
    
    import Text.Printf

    cardValues = [ "empty", "1C", "2C", "3C", "4C", "5C", "6C", "7C", "8C", "9C", "10C", "11C", "12C", "13C", "1D", "2D", "3D", "4D", "5D", "6D", "7D", "8D", "9D", "10D", "11D", "12D", "13D", "1H", "2H", "3H", "4H", "5H", "6H", "7H", "8H", "9H", "10H", "11H", "12H", "13H", "1S", "2S", "3S", "4S", "5S", "6S", "7S", "8S", "9S", "10S", "11S", "12S", "13S" ]

    deal cards = determineWinner (playerOneHand cards) (playerTwoHand cards) cards

    playerOneHand cards = quickSort [ snd x | x <- (zip [0..] cards), fst x /= 1 && fst x /= 3 ]
    playerTwoHand cards = quickSort [ snd x | x <- (zip [0..] cards), fst x /= 0 && fst x /= 2 ]

    quickSort [] = []
    quickSort (x:xs) = quickSort [ y | y <- xs, y `mod` 13 <= x `mod` 13 ] ++ [x] ++ quickSort [ y | y <- xs, y `mod` 13 > x `mod` 13 ]

    determineWinner p1Hand p2Hand cards
        | valueOne > valueTwo = translate comboOne
        | valueOne < valueTwo = translate comboTwo
        | otherwise = translate (kicker valueOne comboOne comboTwo cards)
        where 
            valueOne = fst (evaluateHand p1Hand)
            comboOne = snd (evaluateHand p1Hand)
            valueTwo = fst (evaluateHand p2Hand)
            comboTwo = snd (evaluateHand p2Hand)
    
    translate hand = [ cardValues !! x | x <- hand ]

    evaluateHand hand 
        | fst royalFlushHand    = (10, snd royalFlushHand)
        | fst straightFlushHand = (09, snd straightFlushHand)
        | fst fourOfAKindHand   = (08, snd fourOfAKindHand)
        | fst fullHouseHand     = (07, snd fullHouseHand)
        | fst flushHand         = (06, snd flushHand)
        | fst straightHand      = (05, snd straightHand)
        | fst tripleHand        = (04, snd tripleHand)
        | fst twoPairHand       = (03, snd twoPairHand)
        | fst pairHand          = (02, snd pairHand)
        | otherwise             = (01, getHighCard hand)
        where 
            royalFlushHand    = royalFlushCheck hand
            straightFlushHand = straightFlushCheck hand
            fourOfAKindHand   = duplicateCheck hand 4 (-1)
            fullHouseHand     = fullHouseCheck hand
            flushHand         = flushCheck hand
            straightHand      = straightCheck hand
            tripleHand        = duplicateCheck hand 3 (-1)
            twoPairHand       = twoPairCheck hand
            pairHand          = duplicateCheck hand 2 (-1)

    royalFlushCheck hand
        | clubs    == [ x | x <- hand, x `elem` clubs    ] = (True, clubs)
        | diamonds == [ x | x <- hand, x `elem` diamonds ] = (True, diamonds)
        | hearts   == [ x | x <- hand, x `elem` hearts   ] = (True, hearts)
        | spades   == [ x | x <- hand, x `elem` spades   ] = (True, spades)
        | otherwise = (False, [])
        where 
            clubs    = [13, 01, 10, 11, 12]
            diamonds = [26, 14, 23, 24, 25]
            hearts   = [39, 27, 36, 37, 38]
            spades   = [52, 40, 49, 50, 51]

    straightFlushCheck hand
        | c + 1 == d && c + 2 == e && c + 3 == f && c + 4 == g = (True, [ c, d, e, f, g ])
        | b + 1 == c && b + 2 == d && b + 3 == e && b + 4 == f = (True, [ b, c, d, e, f ])
        | a + 1 == b && a + 2 == c && a + 3 == d && a + 4 == e = (True, [ a, b, c, d, e ])
        | otherwise = (False, [])
        where 
            (a, b, c, d, e, f, g) = convertTuple $ quickSortKing hand
              
    convertTuple [a, b, c, d, e, f, g] = (a, b, c, d, e, f, g)

    duplicateCheck hand duplicateType pair
        | length ace   == duplicateType && pair /= 01 = (True, ace)
        | length king  == duplicateType && pair /= 00 = (True, king)
        | length queen == duplicateType && pair /= 12 = (True, queen)
        | length jack  == duplicateType && pair /= 11 = (True, jack)
        | length ten   == duplicateType && pair /= 10 = (True, ten)
        | length nine  == duplicateType && pair /= 09 = (True, nine)
        | length eight == duplicateType && pair /= 09 = (True, eight)
        | length seven == duplicateType && pair /= 07 = (True, seven)
        | length six   == duplicateType && pair /= 06 = (True, six)
        | length five  == duplicateType && pair /= 05 = (True, five)
        | length four  == duplicateType && pair /= 04 = (True, four)
        | length three == duplicateType && pair /= 03 = (True, three)
        | length two   == duplicateType && pair /= 02 = (True, two)
        | otherwise = (False, [])
        where
            ace   = [ x | x <- hand, x `mod` 13 == 01 ]
            king  = [ x | x <- hand, x `mod` 13 == 00 ]
            queen = [ x | x <- hand, x `mod` 13 == 12 ]
            jack  = [ x | x <- hand, x `mod` 13 == 11 ]
            ten   = [ x | x <- hand, x `mod` 13 == 10 ]
            nine  = [ x | x <- hand, x `mod` 13 == 09 ]
            eight = [ x | x <- hand, x `mod` 13 == 08 ]
            seven = [ x | x <- hand, x `mod` 13 == 07 ]
            six   = [ x | x <- hand, x `mod` 13 == 06 ]
            five  = [ x | x <- hand, x `mod` 13 == 05 ]
            four  = [ x | x <- hand, x `mod` 13 == 04 ]
            three = [ x | x <- hand, x `mod` 13 == 03 ]
            two   = [ x | x <- hand, x `mod` 13 == 02 ]

    fullHouseCheck hand
        | fst pair && fst triple = (True, concat[ snd pair, snd triple ])
        | otherwise = (False, [])
        where 
            pair   = duplicateCheck hand 2 (-1)
            triple = duplicateCheck hand 3 (-1)

    flushCheck hand
        | length clubs    >= 5 = (True, getFlush $ quickSortRank clubs)
        | length diamonds >= 5 = (True, getFlush $ quickSortRank diamonds)
        | length hearts   >= 5 = (True, getFlush $ quickSortRank hearts)
        | length spades   >= 5 = (True, getFlush $ quickSortRank spades)
        | otherwise = (False, [])
        where 
            clubs    = [ x | x <- hand, x `elem` [01..13] ]
            diamonds = [ x | x <- hand, x `elem` [14..26] ]
            hearts   = [ x | x <- hand, x `elem` [27..39] ]
            spades   = [ x | x <- hand, x `elem` [40..52] ]

    quickSortRank hand
        | head hand `mod` 13 == 0 = quickSortRank (tail hand ++ [ head hand ])
        | head hand `mod` 13 == 1 = quickSortRank (tail hand ++ [ head hand ])
        | otherwise = hand

    getFlush hand
        | length hand > 5 = getFlush (tail hand)
        | otherwise = hand

    straightCheck hand
        | fst d + 1 == fst e && fst d + 2 == fst f && fst d + 3 == fst g && fst d - fst a == fst d - 1 && fst g == 13 = (True, [ snd d, snd e, snd f, snd g, snd a ])
        | fst c + 1 == fst d && fst c + 2 == fst e && fst c + 3 == fst f && fst c - fst a == fst c - 1 && fst f == 13 = (True, [ snd c, snd d, snd e, snd f, snd a ])
        | fst b + 1 == fst c && fst b + 2 == fst d && fst b + 3 == fst e && fst b - fst a == fst b - 1 && fst e == 13 = (True, [ snd b, snd c, snd d, snd e, snd a ])
        | fst c + 1 == fst d && fst c + 2 == fst e && fst c + 3 == fst f && fst c + 4 == fst g                        = (True, [ snd c, snd d, snd e, snd f, snd g ])
        | fst b + 1 == fst c && fst b + 2 == fst d && fst b + 3 == fst e && fst b + 4 == fst f                        = (True, [ snd b, snd c, snd d, snd e, snd f ])
        | fst a + 1 == fst b && fst a + 2 == fst c && fst a + 3 == fst d && fst a + 4 == fst e                        = (True, [ snd a, snd b, snd c, snd d, snd e ])
        | otherwise = (False, [])
        where 
            reducedHand = zip [ if x `mod` 13 == 0 then 13 else x `mod` 13 | x <- (quickSortKing hand) ] (quickSortKing hand)
            noDuplicates = removeDuplicates reducedHand
            (a, b, c, d, e, f, g) = convertTuple $ addEmpty noDuplicates

    quickSortKing hand
        | head hand `mod` 13 == 0 = quickSortKing (tail hand ++ [ head hand ])
        | otherwise = hand

    removeDuplicates [] = []
    removeDuplicates (x:xs) = x:(removeDuplicates (remove x xs))
        where
            remove x [] = []
            remove x (y:ys)
                | fst x == fst y = remove x ys
                | otherwise = y:(remove x ys)

    addEmpty hand
        | length hand < 7 = addEmpty $ hand ++ [(-1, -1)]
        | otherwise = hand

    twoPairCheck hand
        | fst firstPair = (fst secondPair, concat[ snd firstPair, snd secondPair ])
        | otherwise = (False, [])
        where 
            firstPair  = duplicateCheck hand 2 (-1)
            secondPair = duplicateCheck hand 2 ((head $ snd firstPair) `mod` 13)
    
    getHighCard hand = [last $ quickSortRank hand]
    
    kicker value comboOne comboTwo cards
        | (value == 5 || value == 9) = kickerCombo comboOne comboTwo True cards
        | otherwise = kickerCombo comboOne comboTwo False cards
    
    kickerCombo comboOne comboTwo isStraight cards
        | isStraight && (straightComboOne > straightComboTwo) = comboOne
        | isStraight && (straightComboOne < straightComboTwo) = comboTwo
        | not isStraight && (getCardValue reducedComboOne > getCardValue reducedComboTwo) = comboOne
        | not isStraight && (getCardValue reducedComboOne < getCardValue reducedComboTwo) = comboTwo
        | otherwise = kickerWholeHand comboOne comboTwo reducedP1Hand reducedP2Hand
        where
            reducedComboOne = [ if x `mod` 13 == 0 then 13 else x `mod` 13 | x <- comboOne ]
            reducedComboTwo = [ if x `mod` 13 == 0 then 13 else x `mod` 13 | x <- comboTwo ]
            straightComboOne
                | 1 `elem` reducedComboOne && 13 `elem` reducedComboOne = getCardValue reducedComboOne
                | otherwise = sum reducedComboOne
            straightComboTwo
                | 1 `elem` reducedComboTwo && 13 `elem` reducedComboTwo = getCardValue reducedComboTwo
                | otherwise = sum reducedComboTwo
            reducedP1Hand = reverse [ if x `mod` 13 == 0 then 13 else x `mod` 13 | x <- (playerOneHand cards) ]
            reducedP2Hand = reverse [ if x `mod` 13 == 0 then 13 else x `mod` 13 | x <- (playerTwoHand cards) ]

    getCardValue cards
        | length cards > 0 && head cards == 01 = 14 + getCardValue (tail cards)
        | length cards > 0 && head cards /= 01 = head cards + getCardValue (tail cards)
        | otherwise = 0

    kickerWholeHand comboOne comboTwo p1Hand p2Hand
        | length p1Hand > 0 && (head p1Hand > head p2Hand) = comboOne
        | length p1Hand > 0 && (head p1Hand < head p2Hand) = comboTwo
        | length p1Hand > 0 && (head p1Hand == head p2Hand) = kickerWholeHand comboOne comboTwo (tail p1Hand) (tail p2Hand)
        | otherwise = []

-- 160 lines
